op-e2e/faultproofs/cannon_benchmark_test.go

```diff
@@ -0,0 +1,196 @@
+	op_e2e "github.com/ethereum-optimism/optimism/op-e2e"
+	"github.com/ethereum-optimism/optimism/op-e2e/bindings"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/wait"
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/predeploys"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBenchmarkCannon_FPP(t *testing.T) {
+	t.Skip("TODO(client-pod#906): Compare total witness size for assertions against pages allocated by the VM")
+
+	op_e2e.InitParallel(t, op_e2e.UsesCannon)
+	ctx := context.Background()
+	cfg := op_e2e.DefaultSystemConfig(t)
+	// We don't need a verifier - just the sequencer is enough
+	delete(cfg.Nodes, "verifier")
+	// Use a small sequencer window size to avoid test timeout while waiting for empty blocks
+	// But not too small to ensure that our claim and subsequent state change is published
+	cfg.DeployConfig.SequencerWindowSize = 16
+	minTs := hexutil.Uint64(0)
+	cfg.DeployConfig.L2GenesisDeltaTimeOffset = &minTs
+	cfg.DeployConfig.L2GenesisEcotoneTimeOffset = &minTs
+
+	sys, err := cfg.Start(t)
+	require.Nil(t, err, "Error starting up system")
+	defer sys.Close()
+
+	log := testlog.Logger(t, log.LevelInfo)
+	log.Info("genesis", "l2", sys.RollupConfig.Genesis.L2, "l1", sys.RollupConfig.Genesis.L1, "l2_time", sys.RollupConfig.Genesis.L2Time)
+
+	l1Client := sys.Clients["l1"]
+	l2Seq := sys.Clients["sequencer"]
+	rollupRPCClient, err := rpc.DialContext(context.Background(), sys.RollupNodes["sequencer"].HTTPEndpoint())
+	require.Nil(t, err)
+	rollupClient := sources.NewRollupClient(client.NewBaseRPCClient(rollupRPCClient))
+	require.NoError(t, wait.ForUnsafeBlock(ctx, rollupClient, 1))
+
+	// Agreed state: 200 Big Contracts deployed at max size - total codesize is 5.90 MB
+	// In Fault Proof: Perform multicalls calling each Big Contract
+	//  - induces 200 oracle.CodeByHash preimage loads
+	// Assertion: Under 2000 pages requested by the program (i.e. max ~8 MB). Assumes derivation overhead; block finalization, etc, requires < 1 MB of program memory.
+
+	const numCreates = 200
+	newContracts := createBigContracts(ctx, t, cfg, l2Seq, cfg.Secrets.Alice, numCreates)
+	receipt := callBigContracts(ctx, t, cfg, l2Seq, cfg.Secrets.Alice, newContracts)
+
+	t.Log("Capture the latest L2 head that preceedes contract creations as agreed starting point")
+	agreedBlock, err := l2Seq.BlockByNumber(ctx, new(big.Int).Sub(receipt.BlockNumber, big.NewInt(1)))
```

Comment:
Don't you really need the first receipt from `createBigContracts` if you want to start before the contract deployments?

Reply:
The `t.Log` message is misleading. We actually want the latest state that precedes the multicall. That is, not including the contract deployments. Otherwise, we'll have to derive multiple L2 blocks which wouldn't be a realistic benchmark given output root bisection).

The optimization for this benchmark is the memory usage / gas - how much memory can induce the op-program to use before we fill a block. We avoid including the contract creations for this reason as they cost a lot of gas. It's more "economical" to just call the big contracts.